crates/extensions/az_mapper_ext/src/converter.rs

```diff
@@ -48,24 +54,89 @@ impl AzureConverter {
 
     fn try_convert(&mut self, input: &MqttMessage) -> Result<Vec<MqttMessage>, ConversionError> {
-        let default_timestamp = self.add_timestamp.then(|| self.clock.now());
+        if is_bridge_health(&input.topic.name) {
+            Ok(vec![])
+        } else {
+            match self.mqtt_schema.entity_channel_of(&input.topic) {
+                Ok((source, channel)) => self.try_convert_te_topics(source, channel, input),
+                Err(_) => self.try_convert_tedge_health_topics(input),
+            }
+        }
+    }
+
+    fn try_convert_tedge_health_topics(
+        &mut self,
+        input: &MqttMessage,
+    ) -> Result<Vec<MqttMessage>, ConversionError> {
         if is_bridge_health(&input.topic.name) {
             Ok(vec![])
         } else if input.topic.name.starts_with("tedge/health") {
             let mut input_msg = input.clone();
-            input_msg.topic = self.mapper_config.out_topic.clone();
+            input_msg.topic = Topic::new_unchecked(&self.mapper_config.out_topic.clone());
             Ok(vec![input_msg])
         } else {
-            self.size_threshold.validate(input)?;
+            Ok(vec![])
+        }
+    }
+
+    fn try_convert_te_topics(
+        &mut self,
+        source: EntityTopicId,
+        channel: Channel,
+        input: &MqttMessage,
+    ) -> Result<Vec<MqttMessage>, ConversionError> {
+        let mea_type = match &channel {
+            Channel::Measurement { measurement_type } => measurement_type,
+            Channel::Event { event_type } => event_type,
+            Channel::Alarm { alarm_type } => alarm_type,
+            _ => return Ok(vec![]),
+        };
+        self.convert_message(input, source, mea_type)
+    }
+
+    fn convert_message(
+        &mut self,
+        input: &MqttMessage,
+        source: EntityTopicId,
+        telemetry_type: &String,
+    ) -> Result<Vec<MqttMessage>, ConversionError> {
+        let payload = self.with_timestamp(input)?;
+        let source = normalize_name(source);
+
+        let out_topic = match input.topic.name.split('/').collect::<Vec<_>>()[..] {
+            [_, _, _, _, _, "m", _] => Topic::new_unchecked(&format!(
+                "{0}{source}/m/{telemetry_type}",
+                self.mapper_config.out_topic
+            )),
+            [_, _, _, _, _, "e", _] => Topic::new_unchecked(&format!(
+                "{}{source}/e/{telemetry_type}",
+                self.mapper_config.out_topic
+            )),
+            [_, _, _, _, _, "a", _] => Topic::new_unchecked(&format!(
+                "{}{source}/a/{telemetry_type}",
+                self.mapper_config.out_topic
+            )),
+            _ => return Ok(vec![]),
+        };
 
-            let mut serializer = ThinEdgeJsonSerializer::new_with_timestamp(default_timestamp);
-            tedge_api::parser::parse_str(input.payload_str()?, &mut serializer)?;
+        let output = MqttMessage::new(&out_topic, payload);
+        self.size_threshold.validate(&output)?;
```

Comment:
In the earlier code, the size validation was happening up-front as well, to avoid unnecessary conversion, if the input is already too large.

Reply:
added